rules/klayout/drc/testing/run_switch_checking.py

```diff
@@ -69,61 +70,103 @@ def run_switches():
-            os.system('echo Test case passed. >> {0}{1}'.format(RUN_DIRECTORY,LOG_FILE_NAME))
+        if len(out) > 0 and expected_result == PASS:
+            logging.info("## Found the expected log: " + str(out))
+            logging.info("Test case passed. {}".format(test_case))
+            logging.info(f'echo Test case passed. >> {RUN_DIRECTORY}{LOG_FILE_NAME}')
+            return True
+        elif len(out) < 1 and expected_result == PASS:
+            logging.error("Test case passed as expected but didn't generate the expected log: {}".format(test_case))
+            return False
         else:
-            logging.error("Test case FAILD check log file for more info.")
-            os.system('echo Test case FAILD check log file for more info. >> {0}{1}'.format(RUN_DIRECTORY,LOG_FILE_NAME))
+            logging.error("Test case FAILD check log file for more info. {}".format(test_case))
+            logging.error(f'echo Test case FAILD check log file for more info. >> {RUN_DIRECTORY}{LOG_FILE_NAME}')
+            return False
 
     ## read testcases from csv file
     df = pd.read_csv(csv_path)
+    pd.set_option('display.max_columns', None)
+    pd.set_option('display.max_rows', None)
+    pd.set_option("max_colwidth", None)
+    pd.set_option('display.width', 1000)
+
+    print("## All Switch checking patterns:")
+    print(df)
+
     test_cases = df["test_case_name"]
     switches = df["switches"]
     expected_results = df["expected_output"]
     expected_log = df["expected_logs"]
 
     ## init log file
-    os.system('mkdir -p {0}'.format(RUN_DIRECTORY))
-    os.system('touch {0}temp.log'.format(RUN_DIRECTORY))
-    os.system('mv -f {0}temp.log {0}{1}'.format(RUN_DIRECTORY,LOG_FILE_NAME))
+    check_call(
+        f'''
+mkdir -p {RUN_DIRECTORY}
+touch {RUN_DIRECTORY}temp.log
+mv -f {RUN_DIRECTORY}temp.log {RUN_DIRECTORY}{LOG_FILE_NAME}
+''',
+        shell=True
+    )
+
+    run_status = []
+    sw_failed = False
 
     for test_case_index in range(0,len(test_cases)):
         test_case_switches = switches[test_case_index]
 
         running_msg = "\nrunning: "+test_cases[test_case_index] + "     with switches: " + test_case_switches + "..."
         logging.info(running_msg)
-        os.system('echo "{0}" >> {1}{2}'.format(running_msg, RUN_DIRECTORY, LOG_FILE_NAME))
+        check_call(f'echo "{running_msg}" >> {RUN_DIRECTORY}{LOG_FILE_NAME}', shell=True)
+
+        case_res = run_test_case(test_cases[test_case_index], test_case_switches, expected_results[test_case_index],expected_log[test_case_index])
+        run_status.append(case_res)
 
-        run_test_case(test_cases[test_case_index], test_case_switches, expected_results[test_case_index],expected_log[test_case_index])
+        if not case_res:
+            sw_failed = True
 
+    df["run_status"] = run_status
+    
+    if sw_failed:
+        logging.error("## One of the test cases failed. Exit with failure:")
+        print(df)
+        exit(1)
```

Comment:
`Exception`?

Reply:
@proppy What's the difference? This will send a exit code of 1 which will fail github actions.

Reply:
Yes, but an Exception give allow you to report more structured error and allow you to potentially handle it in the caller frames, i.e: imagine you have a test runner that run those test functions, you don't want the test itself to `exit(1)` (that would ends up the runner too as they are part of the same process) but you want to runner to be able to catch the `Exception` from the test and proceed accordingly.

Reply:
@proppy But this is not a unit function. We only 'exit(1)' after the entire run is complete.